runtime/src/crowdfund.rs

```diff
@@ -0,0 +1,1104 @@
+
+			// Avoid using transfer to ensure we don't pay any fees.
+			let _ = T::Currency::resolve_into_existing(&who, T::Currency::withdraw(
+				&Self::fund_account_id(index),
+				balance,
+				WithdrawReason::Transfer,
+				ExistenceRequirement::AllowDeath
+			)?);
+
+			Self::contribution_kill(index, &who);
+			fund.raised = fund.raised.saturating_sub(balance);
+
+			<Funds<T>>::insert(index, &fund);
+
+			Self::deposit_event(RawEvent::Withdrew(who, index, balance));
+		}
+		
+		/// Remove a fund after either: it was unsuccessful and it timed out; or it was successful
+		/// but it has been retired from its parachain slot. This places any deposits that were not
+		/// withdrawn into the treasury.
+		fn dissolve(origin, #[compact] index: FundIndex) {
+			let _ = ensure_signed(origin)?;
+
+			let fund = Self::funds(index).ok_or("invalid fund index")?;
+			ensure!(fund.parachain.is_none(), "cannot dissolve fund with active parachain");
+			let now = <system::Module<T>>::block_number();
+			ensure!(now >= fund.end + T::RetirementPeriod::get(), "retirement period not over");
+
+			let account = Self::fund_account_id(index);
+
+			// Avoid using transfer to ensure we don't pay any fees.
+			let _ = T::Currency::resolve_into_existing(&fund.owner, T::Currency::withdraw(
+				&account,
+				fund.deposit,
+				WithdrawReason::Transfer,
+				ExistenceRequirement::AllowDeath
+			)?);
+
+			T::OrphanedFunds::on_unbalanced(T::Currency::withdraw(
+				&account,
+				fund.raised,
+				WithdrawReason::Transfer,
+				ExistenceRequirement::AllowDeath
+			)?);
+
+			Self::crowdfund_kill(index);
+			<Funds<T>>::remove(index);
+
+			Self::deposit_event(RawEvent::Dissolved(index));
+		}
+		
+		fn on_finalize(n: T::BlockNumber) {
+			if <slots::Module<T>>::is_ending(n).is_some() {
+				for (fund, index) in NewRaise::take().into_iter().filter_map(|i| Self::funds(i).map(|f| (f, i)))
+				{
+					let bidder = slots::Bidder::New(slots::NewBidder {
+						who: Self::fund_account_id(index),
+						/// FundIndex and slots::SubId happen to be the same type (u32). If this
+						/// ever changes, then some sort of conversion will be needed here.
+						sub: index,
+					});
+
+					// Care needs to be taken by the crowdfund creator that this function will succeed given
+					// the crowdfunding configuration. We do some checks ahead of time in crowdfund `create`.
+					let _ = <slots::Module<T>>::handle_bid(
+						bidder,
+						<slots::Module<T>>::auction_counter(),
+						fund.first_slot,
+						fund.last_slot,
+						fund.raised,
+					);
+				}
+			}
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// The account ID of the fund pot.
+	///
+	/// This actually does computation. If you need to keep using it, then make sure you cache the
+	/// value and only call this once.
+	pub fn fund_account_id(index: FundIndex) -> T::AccountId {
+		MODULE_ID.into_sub_account(index)
+	}
+
+	pub fn id_from_index(index: FundIndex) -> Vec<u8> {
```

Comment:
why not using `double_map`?

Reply:
Gav's comment in the docs was:

```
//! Contributing funds is permissionless. Each fund has a child-trie which stores all
//! contributors account IDs together with the amount they contributed; the root of this can then be
//! used by the parachain to allow contributors to prove that they made some particular contribution
//! to the project (e.g. to be rewarded through some token or badge). The trie is retained for later
//! (efficient) redistribution back to the contributors.
```

I too had this question, but when talking with @thiolliere, he mentioned that this would isolate the crowdfund tracking trie from the shared module storage trie. Thus, the root of the trie would only represent contributions to that crowdfund. Will let @gavofyork complete the story

Reply:
I see. It will be very useful if this can be generalized later in Substrate as I can see many different usages for this pattern.

Reply:
I think this PR is supposed to help generalize the usage of child tries: https://github.com/paritytech/substrate/pull/2209

Maybe you mean something more though?

Reply:
it is a start point, but not a complete developer friendly solution. e.g. no metadata integration and hence hard JS SDK integration.
I am looking for something like
```
decl_child_storage! {
  ParentKey: FundIndex;
  Contribution: map AccountId => Balances;
}
```

Reply:
yes having child storage stuff into metadata seems important, I'm not aware of any work on that currently, but I think it is important as well.
The PR 2209 should just change a bit the usage of the child trie.

Reply:
https://github.com/paritytech/substrate/issues/3187